virtual_ecosystem/core/core_components.py

```diff
@@ -218,33 +272,120 @@ def __post_init__(self, config: Config) -> None:
 
         # Validate configuration
         self.canopy_layers = _validate_positive_integer(lyr_config["canopy_layers"])
-        self.soil_layers = _validate_soil_layers(lyr_config["soil_layers"])
+        self.soil_layers = np.array(_validate_soil_layers(lyr_config["soil_layers"]))
 
         # Other heights should all be positive floats
         for attr, value in (
             ("above_canopy_height_offset", lyr_config["above_canopy_height_offset"]),
             ("surface_layer_height", lyr_config["surface_layer_height"]),
-            ("subcanopy_layer_height", lyr_config["subcanopy_layer_height"]),
         ):
             setattr(self, attr, _validate_positive_finite_numeric(value, attr))
 
-        self.layer_roles = (
+        # Get the layer role sequence
+        self.layer_roles = np.array(
             ["above"]
             + ["canopy"] * int(self.canopy_layers)
-            + ["subcanopy"]
             + ["surface"]
-            + ["soil"] * len(self.soil_layers)
+            + ["topsoil"]
+            + ["subsoil"] * (len(self.soil_layers) - 1)
         )
 
+        # Record the number of layers and layer indices
         self.n_layers = len(self.layer_roles)
+        self.layer_indices = np.arange(0, self.n_layers)
+
+        # Set up the various indices onto those roles
+        self.role_indices_bool = {
+            layer_role: self.layer_roles == layer_role
+            for layer_role in ("above", "canopy", "surface", "topsoil", "subsoil")
+        }
+        self.role_indices = {
+            ky: np.nonzero(vl)[0] for ky, vl in self.role_indices_bool.items()
+        }
```

Comment:
If you want to go a step further with the syntactic sugar, you could create properties for all of these. It is more code here, but will reduce the verbosity in other parts of the codebase as well as save them from typos as people will not need to use a string to access the `role_indices` dictionaries:

```python
@property 
def above_index(self):
    return self.role_indices["above"]

@property 
def above_bool(self):
    return self.role_indices_bool["above"]
...
```
As I said, this is just syntactic sugar, but I think it will have a positive impact in readability.

Reply:
I did consider that and I agree this would be more readable. It doesn't work so cleanly with the user-requested aggregate indices over in #452.

Reply:
@dalonsoa Do you have any thoughts on whether integer or boolean indexing is likely to be more efficient? We'll be doing it a lot. We don't _need_ both boolean and integer indices, I think, so if one is clearly better we can simplify.